### PR TITLE
Re-write readme

### PR DESCRIPTION
**Pull request by paulhauner**
This is a full re-write of the reamde, aiming to encourage people to reach out and help contribute to the library.

It also provides an introduction to Ethereum 2.0

The audience is a technical audience familiar with Ethereum.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,59 +1,245 @@
...
+Ethereum 2.0 is not fully specified and there's no working implementation. Some
+teams have demos available, however these demos indicate _progress_, not
+completion. We look forward to providing user functionality once we are ready
+to provide a minimum-viable user experience.
+
+The work-in-progress specification lives
+[here](https://github.com/ethereum/eth2.0-specs/blob/master/specs/casper_sharding_v2.1.md)
+in the [ethereum/eth2.0-specs](https://github.com/ethereum/eth2.0-specs)
+repository. It is still in a draft phase, however there are several teams
+already implementing it. This spec is being developed by Ethereum Foundation
```
Comment (peanutFactory) [inline README.md]: The specification is still in draft phase, however, there are several teams
 already implementing it.

- initial "It" was ambiguous

File: README.md [hunk 2]
```diff
@@ -1,59 +1,245 @@
...
+researchers, most notably Vitalik Buterin. There is active discussion about the
+spec in the [ethereum/sharding](https://gitter.im/ethereum/sharding) gitter
+channel. There is a proof-of-concept implementation in Python at
+[ethereum/beacon_chain](https://github.com/ethereum/beacon_chain).
```
Comment (peanutFactory) [inline README.md]: A proof-of-concept implementation in Python is available at
 [ethereum/beacon_chain](https://github.com/ethereum/beacon_chain).

- two consecutive sentences started with "There"

File: README.md [hunk 3]
```diff
@@ -1,59 +1,245 @@
...
+
+Presently, the spec almost exclusively defines the Beacon Chain as it
+is the focus of present development efforts. Progress on shard chain
```
Comment (peanutFactory) [inline README.md]: Presently, the spec almost exclusively defines the Beacon Chain, which
 is the focus of present development efforts.

File: README.md [hunk 4]
```diff
@@ -1,59 +1,245 @@
...
+### Goals
+
+We aim to contribute to the research and development of a secure, efficient and
+decentralised Ethereum protocol through the development of an open-source
+Ethereum 2.0 client.
+
+We aim to provide a secure and efficient open-source client. As well as
+building an implementation, we seek to help maintain and improve the protocol
+wherever possible.
+
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Accepted

File: README.md [hunk 5]
```diff
@@ -1,59 +1,245 @@
...
+### Components
+
+Lighthouse is presently focussed on the Beacon Chain implementation. Here are
+some of the components actively under development by the team:
+
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Accepted

File: README.md [hunk 6]
```diff
@@ -1,59 +1,245 @@
...
+- **BLS cryptography**: we presently use the [Apache
+  Milagro](https://milagro.apache.org/) cryptography library to create and
+verify BLS aggregate signatures. These signatures are core to Eth 2.0 and allow
+the signatures of many validators to be compressed into a constant 96 bytes and
+verified efficiently. We're presently maintaining our own [BLS aggregates
+library](https://github.com/sigp/signature-schemes), gratefully forked from
+@lovesh.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Accepted

File: README.md [hunk 7]
```diff
@@ -1,59 +1,245 @@
...
+- **DoS-resistant block pre-processing**: processing blocks in proof-of-stake
+  is more resource intensive than proof-of-work. As such, clients need to
+ensure that bad blocks can be rejected as efficiently as possible. We can
+presently process a block with 10 million ETH staked in 0.006 seconds and
+reject valid blocks even quicker. See the
+[issue](https://github.com/ethereum/beacon_chain/issues/103) on
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Should have been invalid, thanks!

File: README.md [hunk 8]
```diff
@@ -1,59 +1,245 @@
...
+[ethereum/beacon_chain](https://github.com/ethereum/beacon_chain)
+.
+- **P2P networking**: Eth 2.0 is likely use the [libp2p
+  framework](https://libp2p.io/), lighthouse hopes to work alongside
+[Parity](https://www.parity.io/) to get
+[libp2p-rust](https://github.com/libp2p/rust-libp2p) fit-for-purpose.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Accepted

File: README.md [hunk 9]
```diff
@@ -1,59 +1,245 @@
...
+- **Validator duties** : providing "validator" services to users who wish to
+  stake ETH. This involves maintaining a consistent view of the chain and
+voting upon shard and beacon chain blocks.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Yeah, my description wasn't great. Your understanding was correct and I have used it :)

File: README.md [hunk 10]
```diff
@@ -1,59 +1,245 @@
...
+- **New serialization formats**: lighthouse is working alongside the EF
+  researchers to develop "simpleserialize" a purpose-built serialization format
+for sending information across the network. Check our our [SSZ
+implementation](https://github.com/sigp/lighthouse/tree/master/beacon_chain/utils/ssz)
+and our
+[research](https://github.com/sigp/serialization_sandbox/blob/report/report/serialization_report.md)
+on serialization formats.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Accepted

File: README.md [hunk 11]
```diff
@@ -1,59 +1,245 @@
...
+- **Casper FFG fork-choice**: the [Casper
+  FFG](https://arxiv.org/abs/1710.09437) fork-choice rules allow the chain to
+select a canonical chain in the case of a fork.
+- **Efficient state transition logic**: "state transition" logic deals with
+  updating the validator set as valiators log in/out, penalising/rewarding
+validators, rotating validators across shards, and many other core tasks.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Accepted

File: README.md [hunk 12]
```diff
@@ -1,59 +1,245 @@
...
+- **Fuzzing and testing environments**: we preparing to implement lab
+  environments with CI work-flows to provide automated security analysis.
+
+On top of these components we're also working on database schemas,
+RPC frameworks, specification development, database optimizations (e.g.,
+bloom-filters) and tons of other interesting stuff (at least we think so).
+
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Thank you! Accepted

File: README.md [hunk 13]
```diff
@@ -1,59 +1,245 @@
...
+[research](https://github.com/sigp/serialization_sandbox/blob/report/report/serialization_report.md)
+on serialization formats.
+- **Casper FFG fork-choice**: the [Casper
+  FFG](https://arxiv.org/abs/1710.09437) fork-choice rules allow the chain to
+select a canonical chain in the case of a fork.
+- **Efficient state transition logic**: "state transition" logic deals with
+  updating the validator set as valiators log in/out, penalising/rewarding
+validators, rotating validators across shards, and many other core tasks.
+- **Fuzzing and testing environments**: we preparing to implement lab
+  environments with CI work-flows to provide automated security analysis.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Accepted

File: README.md [hunk 14]
```diff
@@ -1,59 +1,245 @@
...
+
+If you would like to learn more about Ethereum 2.0 and/or
+[Rust](https://www.rust-lang.org/), we would be more than happy to on-board you
+and assign you to some tasks. We aim to be as accepting and understanding as
+possible; we are more than happy to up-skill contributors in exchange for their
+help on the project.
+
+Alternatively, if you an ETH/Rust veteran we'd love to have your input.  We're
+always looking for the best way to do things and will always consider any
+respectfully presented criticism.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Accepted

File: README.md [hunk 15]
```diff
@@ -1,59 +1,245 @@
...
+**NOTE: the cryptography libraries used in this implementation are
+experimental and as such all cryptography should be assumed to be insecure.**
 
-It is an implementation of the [Full PoS Casper chain
-v2](https://notes.ethereum.org/SCIg8AH5SA-O4C1G1LYZHQ?view) spec and is also
-largely based upon the
-[ethereum/beacon_chain](https://github.com/ethereum/beacon_chain) repo.
+The code-base is still under-development and does not provide any user-facing
+functionality. However, there are a lot of tests and some benchmarks if that
+sort of thing interests you.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Great point. I have changed this to:

"For developers and researchers, there are tests and benchmarks which could be of interest."

File: README.md [hunk 16]
```diff
@@ -1,59 +1,245 @@
...
-## Progress
+Generally, tests can be kept in the same file, as is typical in Rust.
+Integration tests should be placed in the `tests` directory in the crates root.
+Particularity large (line-count) tests should be separated into another file.
 
-As of 02/08/2018, there is a basic libp2p implementation alongside a series of
-state objects and state transition functions. There are no syncing capabilities.
+A function is not complete until it is tested. We produce tests to help
+ourselves and others understand others code and to provide protection from
+regression (breaking stuff accidentally).
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Thank you, this is very ugly.

I've changed it to "We produce tests to protect against regression (accidentally breaking things) and to help those who read out code to understand how the function should (or shouldn't) be used."

File: README.md [hunk 17]
```diff
@@ -1,59 +1,245 @@
...
+The best place for discussion is the [sigp/lighthouse](https://gitter.im/sigp/lighthouse) gitter.
+Ping @paulhauner or @AgeManning to get the quickest response.
+
+
+# What is Ethereum 2.0
+
+Ethereum 2.0 is the name that has been given to a new blockchain being
+developed by the Ethereum Foundation and the Ethereum community. Ethereum 2.0
+consists of 1,025 proof-of-stake blockchains; the "beacon chain" and 1,024
+"shard chains".
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Much better, thanks

File: README.md [hunk 18]
```diff
@@ -1,59 +1,245 @@
...
+
+## Beacon Chain
+
+The Beacon Chain is a little different to more common blockchains like Bitcoin
+and present-Ethereum in that it doesn't process "transactions", per say.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Great, thanks

File: README.md [hunk 19]
```diff
@@ -1,59 +1,245 @@
...
-chain. 
+```
+$ cargo bench --all
+```
 
-Whilst the Beacon Chain relies upon the PoW chain for block hashes, Lighthouse
-will need to run alongside an existing client (e.g., Geth, Parity Ethereum),
-only being able to stand by itself once the PoW chain has been deprecated.
+Lighthouse presently runs on Rust `stable`, however benchmarks require the
+`nightly` version.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Added. thanks

File: README.md [hunk 20]
```diff
@@ -1,59 +1,245 @@
...
+Ethereum 2.0 is the name that has been given to a new blockchain being
+developed by the Ethereum Foundation and the Ethereum community. Ethereum 2.0
+consists of 1,025 proof-of-stake blockchains; the "beacon chain" and 1,024
+"shard chains".
+
+## Beacon Chain
+
+The Beacon Chain is a little different to more common blockchains like Bitcoin
+and present-Ethereum in that it doesn't process "transactions", per say.
+Instead, it maintains a set of bonded (staked) validators and co-ordinates them
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Great!

File: README.md [hunk 21]
```diff
@@ -1,59 +1,245 @@
...
+to provide services to a static set of "sub-blockchains" (shards). These shards
+then process the normal "5 ETH from A to B" transactions in _parallel_ whilst
+deferring consensus to the Beacon Chain.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Great

File: README.md [hunk 22]
```diff
@@ -1,59 +1,245 @@
...
+
+Here are some of the major services that the beacon chain provides to its
+shards:
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Excellent (I feel like I'm over doing it by replying to _all_ of these comments, but I've started now)

File: README.md [hunk 23]
```diff
@@ -1,59 +1,245 @@
...
+- Valdidator management, including:
+    - Inducting and ejecting validators.
+    - Delegating randomly-shuffled subsets of validators to validate shards.
+    - Penalising and rewarding validators.
+- Proof-of-stake consensus for shard chain blocks.
+
+## Shard Chains
+
+Shards can be thought of like CPU cores, they're a lane where transactions can
+execute in series (one-after-another). Presently, Ethereum is single-core and
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: In a wild turn of events, I have accepted and implemented this suggestion.

File: README.md [hunk 24]
```diff
@@ -1,59 +1,245 @@
...
+can only _fully_ process one transaction at a time. Sharding allows multiple
+transactions to happen in parallel, greatly increasing the per-second
+transaction capacity of Ethereum.
+
+Each shard is proof-of-stake and shares its validators (stakers) with the other
+shards as the beacon chain rotates validators pseudo-randomly across shards.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: I think "uses" works best here

File: README.md [hunk 25]
```diff
@@ -1,59 +1,245 @@
...
+Shards will likely be the basis of very interesting layer-2 transaction
+processing schemes, however we won't get into that here.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Bueno

File: README.md [hunk 26]
```diff
@@ -1,59 +1,245 @@
...
+
+## The Proof-of-Work Chain
+
+The proof-of-work chain will hold a contract to allow addresses to deposit 32
+ETH, a BLS public key and some [other
+parameters](https://github.com/ethereum/eth2.0-specs/blob/master/specs/casper_sharding_v2.1.md#pow-chain-changes)
+to allow them to become Beacon Chain validators. Each Beacon Chain will
+reference a PoW block hash allowing PoW clients to use the Beacon Chain as a
+source of [Casper FFG finality](https://arxiv.org/abs/1710.09437), if they
+desire.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Molto bene

File: README.md [hunk 27]
```diff
@@ -1,59 +1,245 @@
...
+shards as the beacon chain rotates validators pseudo-randomly across shards.
+Shards will likely be the basis of very interesting layer-2 transaction
+processing schemes, however we won't get into that here.
+
+## The Proof-of-Work Chain
+
+The proof-of-work chain will hold a contract to allow addresses to deposit 32
+ETH, a BLS public key and some [other
+parameters](https://github.com/ethereum/eth2.0-specs/blob/master/specs/casper_sharding_v2.1.md#pow-chain-changes)
+to allow them to become Beacon Chain validators. Each Beacon Chain will
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: !detpeccA

File: README.md (hunk 1 above)
Comment (paulhauner) [reply README.md]: Had to reword a bunch of this, so I fixed this too :)

File: README.md [hunk 28]
```diff
@@ -1,59 +1,245 @@
...
+reference a PoW block hash allowing PoW clients to use the Beacon Chain as a
+source of [Casper FFG finality](https://arxiv.org/abs/1710.09437), if they
+desire.
+
+## Ethereum 2.0 Progress
+
+Ethereum 2.0 is not fully specified and there's no working implementation. Some
+teams have demos available, however these demos indicate _progress_, not
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply README.md]: Accepted!

File: README.md (hunk 2 above)
Comment (paulhauner) [reply README.md]: Nice

File: README.md (hunk 3 above)
Comment (paulhauner) [reply README.md]: Accepted!

**Comment by paulhauner** [comment]
Thank you @peanutFactory! So many bad sentences, so many good suggestions!

**Comment by paulhauner** [comment]
Alrighty, after reviews from @peanutFactory and @zedt3ster I am merging this into master. If you have any other comments, make an issue :)

**Pull request merged**